CMSgov/mint-app
# Easi 1978/bsg questions

# EASI-1978

## Changes and Description

- Introduce Plan_Discussion
    - GQL
        -  Get collection
        - Create
        - Update
        - Delete
     - Unit Tests
     - DevData
- Introduce Discussion_Reply
   - GQL
        -  Get collection
        - Create
        - Update
        - Delete
   - Unit Tests
   - DevData


## How to test this change
1. run unit tests
2. start backend, run db:clean && db:seed

3. To see seeded data run the below query
```gql
query ModelPlan{
  
  modelPlan( id: "6e224030-09d5-46f7-ad04-4bb851b36eab") {
    id
    createdBy
    status
    archived
    collaborators {
      id
      modelPlanID
      euaUserID
      teamRole
      fullName
    }
    discussions {
        id
        modelPlanID
        content
        status
        createdBy
        createdDts
        modifiedBy
        modifiedDts
        replies {
            id
            discussionID
            content
            resolution
            createdBy
            createdDts
            modifiedBy
            modifiedDts

            }
    }
    basics{
        id
        modelPlanID
        modelType
        problem
        goal
        testInventions
        note
        createdBy
        createdDts
        modifiedBy
        modifiedDts
        status        
    }
  }
}
```

4. Create Plan Discussion
```gql
mutation createPlanDiscussion{
#   6e224030-09d5-46f7-ad04-4bb851b36eab --> This is an auto generated block
  createPlanDiscussion(input:{
        id: "0700b5ce-3c07-488f-bfcc-87d9c6fc93fd"
      modelPlanID: "f11eb129-2c80-4080-9440-439cbe1a286f"
      content: "Is this GQL resolver set up correctly"

 } ){
    
id
modelPlanID
  content
  status

createdBy
createdDts
modifiedBy
modifiedDts

  }
  
  
}
```
5. Plan Discussion Update
```gql
mutation updatePlanDiscussion{
#   6e224030-09d5-46f7-ad04-4bb851b36eab --> This is an auto generated block
  updatePlanDiscussion(input:{
        id: "0700b5ce-3c07-488f-bfcc-87d9c6fc93fd"
        modelPlanID: "f11eb129-2c80-4080-9440-439cbe1a286f"
        content: "Is this GQL resolver set up correctly"
        status: UNANSWERED

 } ){
    
id
modelPlanID
  content
  status

createdBy
createdDts
modifiedBy
modifiedDts

  }
  
  
}
```

6. Discussion Reply Create
```gql
mutation createDiscussionReply{
#   6e224030-09d5-46f7-ad04-4bb851b36eab --> This is an auto generated block
  createDiscussionReply(input:{
        id: "d08f3973-336a-4d54-a233-9fd001b2c418"
        discussionID: "0700b5ce-3c07-488f-bfcc-87d9c6fc93fd"
        content: "Let me look into it, ok?"

 } ){
    
    id
    discussionID
    content
    resolution

    createdBy
    createdDts
    modifiedBy
    modifiedDts

  }
  
  
}
```
7. Discussion Reply Update

```gql
mutation updateDiscussionReply{
#   6e224030-09d5-46f7-ad04-4bb851b36eab --> This is an auto generated block
  updateDiscussionReply(input:{
        id: "d08f3973-336a-4d54-a233-9fd001b2c418"
        discussionID: "0700b5ce-3c07-488f-bfcc-87d9c6fc93fd"
        content: "Let me look into it, ok?"
        resolution: true

 } ){
    
    id
    discussionID
    content
    resolution

    createdBy
    createdDts
    modifiedBy
    modifiedDts

  }
  
  
}
```
8. Discussion Reply Delete
```gql
mutation deleteDiscussionReply{
#   6e224030-09d5-46f7-ad04-4bb851b36eab --> This is an auto generated block
  deleteDiscussionReply(input:{
        id: "d08f3973-336a-4d54-a233-9fd001b2c418"
        discussionID: "0700b5ce-3c07-488f-bfcc-87d9c6fc93fd"
        content: "Let me look into it, ok?"
        resolution: true

 } ){
    
    id
    discussionID
    content
    resolution

    createdBy
    createdDts
    modifiedBy
    modifiedDts

  }
  
  
}
```
9. Plan Discussion Delete
```gql
mutation deletePlanDiscussion{
#   6e224030-09d5-46f7-ad04-4bb851b36eab --> This is an auto generated block
  deletePlanDiscussion(input:{
        id: "0700b5ce-3c07-488f-bfcc-87d9c6fc93fd"
        modelPlanID: "f11eb129-2c80-4080-9440-439cbe1a286f"
        content: "Is this GQL resolver set up correctly"
        status: UNANSWERED

 } ){
    
id
modelPlanID
  content
  status

createdBy
createdDts
modifiedBy
modifiedDts

  }
  
  
}
```
NOTE, this is a base implementation. If desired we can address additional business logic. Right now, the assumption is that the front end will implement to suite the current requirements.

Additional possible areas for improvement
- Get all Discussion replies in the DB every time you get a Plan_Discussion
    - If the Front End always gets all the replies, this would be more efficient because otherwise you would get all Discussions, and have to do another query for each Discussion, instead of returning the replies as part of the object
- We could also update the Discussions status based on the Reply objects resolution status on the DB / backend layer. Right now this is not implemented. In order to update the Discussion Status, there needs to be a mutation to manually update it. This could be changed, but this allows greater flexibility in use for the moment.

## PR Author Review Checklist




## PR Reviewer Guidelines
- It's best to pull the branch locally and test it, rather than just looking at the code online!
- Check that all code is adequately covered by tests - if it isn't feel free to suggest the addition of tests.
- Always make comments, even if it's minor, or if you don't understand why something was done.
